src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectConverterFactory.cs

```diff
@@ -99,21 +99,21 @@ public override JsonConverter CreateConverter(Type typeToConvert, JsonSerializer
             return converter;
         }
 
-        private JsonConverter CreateKeyValuePairConverter(Type type, JsonSerializerOptions options)
+        private JsonConverter CreateKeyValuePairConverter(Type type)
         {
             Debug.Assert(type.IsKeyValuePair());
 
             Type keyType = type.GetGenericArguments()[0];
             Type valueType = type.GetGenericArguments()[1];
 
             JsonConverter converter = (JsonConverter)Activator.CreateInstance(
-                typeof(KeyValuePairConverter<,>).MakeGenericType(new Type[] { keyType, valueType }),
+                typeof(SmallObjectWithParameterizedConstructorConverter<,,,,>).MakeGenericType(new Type[] { type, keyType, valueType, JsonTypeInfo.ObjectType, JsonTypeInfo.ObjectType }),
```

Comment:
Why not just make the necessary changes in the `KeyValuePairConverter<,>` class?

Reply:
> Why not just make the necessary changes in the `KeyValuePairConverter<,>` class?
I thought about that and it seemed like KeyValuePairConverter would be just a wrapper for SmallObjectWithParameterizedConstructorConverter. However, I don't mind saving KeyValuePairConverter if it helps to be less complicated.

> If `KeyValuePairConverter<,>` can't be used, shouldn't it be deleted?
I didn't manage to delete it due to some compilation errors, so I ended up showing the implementation first and if everything was ok I would get rid of that.

Reply:
> However, I don't mind saving KeyValuePairConverter if it helps to be less complicated.

Yes please. I don't think picking the appropriate constructor should be the responsibility of the converter factory.